README.md
wikimapR, an R package for importing Wikimapia data as Simple Features
via API
================

<!-- README.md is generated from README.Rmd. Please edit that file -->

<a href="https://zenodo.org/badge/latestdoi/161494897"
target="_blank"><img src="https://zenodo.org/badge/161494897.svg"
alt="DOI" /></a>

`wikimapR` is an R package for accessing the raw vector data from
Wikimapia via official
<a href="http://wikimapia.org/api" target="_blank">Wikimapia API</a>.
Map data is returned as
<a href="https://cran.r-project.org/package=sf" target="_blank">Simple
Features (<code>sf</code>)</a> objects with some of the object details
included as nested `lists`.

WARNING: the package may NOT work for now because of the breaking change
in the Wikimapia API. More details at
<a href="https://github.com/e-kotov/wikimapR/issues/2"
target="_blank">https://github.com/e-kotov/wikimapR/issues/2</a>. Using
‘example’ API key does not seem to work, however everything seems to be
working with private API keys.

**This package is at a VERY alpha stage. Provided ‘as is’. Use with
caution.**

You may also want to try
<a href="https://github.com/plandex/wikimapia-api-py" target="_blank">a
similar package for Python</a>.

### Installation

To install:

``` r
# Install the development version
# install.packages("remotes")
remotes::install_github("e-kotov/wikimapR")
```

To load the package and check the version:

``` r
library(wikimapR)
packageVersion("wikimapR")
#> [1] '0.1.2'
```

Load additional packages:

``` r
library(sf)
library(magrittr)
library(dplyr)
library(purrr)
library(progress)
```

### Usage

Set your API key. There is the API key ‘example’ that’s used by the
testing page <a href="https://wikimapia.org/api/?action=examples"
target="_blank">https://wikimapia.org/api/?action=examples</a>. It is
limited to one request every 30 seconds. Get your own API key at
<a href="https://wikimapia.org/api/?action=my_keys"
target="_blank">https://wikimapia.org/api/?action=my_keys</a> to get up
to 100 requests in 5 minutes.

``` r
# change to your own key
set_wikimapia_api_key("your_key")
# set_wikimapia_api_key("example) # use 'example' key with 1 request per 30 seconds rate limit
```

#### Choose a bbox

Have a look at <a href="https://boundingbox.klokantech.com"
target="_blank">https://boundingbox.klokantech.com</a> and choose a
bounding box. For example: `-1.617122,53.764541,-1.467519,53.831568` for
Leeds and it’s surroundings.

``` r
bbox <- c(-1.669629,53.739816,-1.422465,53.869239) # Leeds
```

#### Get the number of objects in this bounding box

Use <a href="http://wikimapia.org/api#oldbox"
target="_blank"><code>box</code></a> API function to get the number of
features in your area of interest.

``` r
wm <- wm_get_from_bbox(x = bbox, get_location = FALSE, meta_only = TRUE)
wm$found
#> [1] 1320
```

Now we know how many objects we have in the bounding box.

There are `page` and `count` parameters in the Wikimapia API, but you
cannot request objects from a bounding box beyond 10 000 no matter how
high you set the `page` parameter. So you have to subdivide the large
bounding box that you have into smaller bounding boxes with a maximum of
10 000 objects each.

#### Subdivide the bounding box into smaller ones

`subdivide_bbox()` subdivides a large bounding box into smaller ones and
returns `sf polygons`, or `bbox` objects or both. It is good for large
areas with defaults tuned to cities like Moscow.

``` r
small_bboxes <- subdivide_bbox(x = bbox, bbox_cell_size = 0.1, return_bbox_or_sf = "both")
#> 6 bounding boxes created with approximate cell size of 6577x11120 meters.
plot(small_bboxes$sf$geometry)
```

<img src="man/figures/README-subdivide_bbox-1.png" width="672" />

In this example the bounding box cell size is in degrees (need to
somehow fix that to work with meters across the globe). Default of 0.045
degrees is reasonably large, roughly equivalent to 2845x5010 meters. It
has proved to fit \< 10 000 objects per cell in Moscow, where the
density of objects is quite high. For less object-dense cities you may
be able to get away with larger grid of bounding boxes.

This is not ideal and it would be more convenient to create this
subdivision using precise metric, but this will do fine for now.

#### Check then number of objects per cell

Just to be sure that every bounding box that we generated has \<= 10 000
objects, let us query all the bounding boxes. For the current example
with 6 it will take about 3 minutes, as with “example” API key the
cool-down is about 30 seconds.

``` r
# get objects in every bbox, but no need to get the location for now
objects_in_bboxes <- small_bboxes$bbox %>% purrr::map(
  ~wm_get_from_bbox(x = .x, get_location = FALSE), .progress = T )
```

Now we look at the histogram of the number of objects in the small
bounding boxes and the maximum value:

``` r
n_by_bbox <- objects_in_bboxes %>% purrr::map_int(~ .x$meta$found) # extract the number of found objects for every bounding box
max(n_by_bbox)
#> [1] 718
hist(n_by_bbox)
```

<img src="man/figures/README-b_per_bbox-1.png" width="672" />

Since the maximum is well within 10 000, we can proceed to collect the
objects IDs and then download attributes for individual features. We
have to download the detailed objects features one-by-one as
<a href="http://wikimapia.org/api#oldbox"
target="_blank"><code>box</code></a> API only returns object ID, name
and URL. So the strategy for getting all object details is to make a
list of object IDs using <a href="http://wikimapia.org/api#oldbox"
target="_blank"><code>box</code></a> API and then use
<a href="http://wikimapia.org/api#placegetbyid"
target="_blank"><code>place.getbyid</code></a>.

#### Create a list of IDs to fetch

``` r
id_list <- objects_in_bboxes %>%
  purrr::map(~ .x$df$id) %>% # pull object IDs from individual bbox query results data.frames
  unlist() # bind together

str(id_list)
#>  chr [1:461] "1869857" "25258348" "14406815" "25257158" "25259631" ...
head(id_list)
#> [1] "1869857"  "25258348" "14406815" "25257158" "25259631" "14406851"
```

Now we have a list of 461 object IDs that we want to get the details
for.

#### Get detailed data for Wikimapia objects

Let’s take just 3 first objects for this example. It will take up to 1.5
minutes to fetch them with all the details using the ‘example’ API.

``` r
short_list <- id_list[1:3]
wm_objects <- wm_get_by_id(ids = short_list)
```

#### We have the objects!

``` r
str(wm_objects, max.level = 1, nchar.max = 50)
#> Classes 'sf' and 'data.frame':   3 obs. of  25 variables:
#>  $ id                    : int  1869857 25258348 14406815
#>  $ object_type           : int  1 1 1
#>  $ language_id           : int  0 0 0
#>  $ language_iso          : chr  "en" "en" "en"
#>  $ language_name         : chr  "English" "English" "English"
#>  $ urlhtml               : chr  "<a class=\"wikimapia-link\" href="| __truncated__ "<a class=\"wikimapia-link\" href="| __truncated__ "<a class=\"wikimapia-link\" href="| __truncated__
#>  $ title                 : chr  "Pudsey" "Farnley" "Morley"
#>  $ description           : chr  "Pudsey is a market town in West Y"| __truncated__ "District of the city of Leeds." "The small West Yorkshire town of "| __truncated__
#>  $ wikipedia             : chr  "http://en.wikipedia.org/wiki/Pudsey" "http://en.wikipedia.org/wiki/Farnley,_Leeds" "http://en.wikipedia.org/wiki/morley"
#>  $ is_building           : logi  FALSE FALSE FALSE
#>  $ is_region             : logi  FALSE FALSE FALSE
#>  $ is_deleted            : logi  FALSE FALSE FALSE
#>  $ parent_id             : chr  "0" "0" "0"
#>  $ x                     : chr  "-16639084" "-16233262" "-15956783"
#>  $ y                     : chr  "537924232" "537853660" "537441159"
#>  $ pl                    : num  32078 25832 22310
#>  $ is_protected          : logi  FALSE FALSE FALSE
#>  $ user_id               : chr  "904949" "904949" "539863"
#>  $ user_name             : chr  "Timea" "Timea" "jagpalpur"
#>  $ date                  : chr  "1427742742" "1427742836" "1258319330"
#>  $ deletion_state        : chr  "FALSE" "FALSE" "FALSE"
#>  $ is_in_deletion_queue  : chr  "FALSE" "FALSE" "FALSE"
#>  $ is_in_undeletion_queue: chr  "FALSE" "FALSE" "FALSE"
#>  $ details               :List of 3
#>  $ geometry              :sfc_POLYGON of length 3; first list element: List of 1
#>   ..- attr(*, "class")= chr [1:3] "XY" "POLYGON" "sfg"
#>  - attr(*, "sf_column")= chr "geometry"
#>  - attr(*, "agr")= Factor w/ 3 levels "constant","aggregate",..: NA NA NA NA NA NA NA NA NA NA ...
#>   ..- attr(*, "names")= chr [1:24] "id" "object_type" "language_id" "language_iso" ...
```

##### We can now plot them

``` r
plot(wm_objects$geometry)
```

<img src="man/figures/README-obj_plot-1.png" width="672" />

##### We can explore the attributes

``` r
names(wm_objects)
#>  [1] "id"                     "object_type"            "language_id"           
#>  [4] "language_iso"           "language_name"          "urlhtml"               
#>  [7] "title"                  "description"            "wikipedia"             
#> [10] "is_building"            "is_region"              "is_deleted"            
#> [13] "parent_id"              "x"                      "y"                     
#> [16] "pl"                     "is_protected"           "user_id"               
#> [19] "user_name"              "date"                   "deletion_state"        
#> [22] "is_in_deletion_queue"   "is_in_undeletion_queue" "details"               
#> [25] "geometry"
```

##### Most of the details are currently in a nested list for every object

You can use
<a href="https://github.com/tidyverse/purrr" target="_blank">purrr</a>
and/or
<a href="https://github.com/renkun-ken/rlist" target="_blank">rlist</a>
packages to pull any of the details from these nested lists.

``` r
str(wm_objects$details[[1]], max.level = 1)
#> List of 5
#>  $ tags              :List of 1
#>  $ photos            : list()
#>  $ comments          : list()
#>  $ location          :List of 15
#>  $ availableLanguages:List of 1
```

### To-Do List

- Rewrite `subdivide_bbox()` to accept metric values for bbox dimensions
  (not critical, but may be useful for other projects)

- Create a few more helper functions to abstract the user from the calls
  to `purrr` for simple things like getting the number of found objects,
  or for getting number of objects per bounding box (see examples above
  in the Usage section)

- Create a hidden environment variable for storing API key and using it
  automatically in the package functions

- Implement the rest of the Wikimapia API functions

- Fix bugs if any

- Make code more robust

- Write unit tests

- Your suggestions are welcome via GitHub issues for this package

- Submit the package to CRAN someday..?

### Citation

``` r
citation("wikimapR")
print(citation("wikimapR"), bibtex=TRUE)
```

To cite wikimapR in publications, please use:

> Kotov E. (2018). wikimapR: Import Wikimapia Data as Simple Features via
> API. DOI: 10.5281/zenodo.3459878.
> <https://github.com/e-kotov/wikimapR>

A BibTeX entry for LaTeX users is:

      @Manual{,
        title = {wikimapR: Import Wikimapia Data as Simple Features via API},
        author = {Egor Kotov},
        doi = {10.5281/zenodo.3459878},
        year = {2018},
        url = {https://github.com/e-kotov/wikimapR},
      }

### License

The MIT License (MIT) + License File

Copyright © 2018 Egor Kotov

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the
“Software”), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


NEWS.md
# wikimapR 0.1.2 (2023-05-14)

## New features

- Added `set_wikimapia_api_key()` to make API keys available for all Wikimapia API
calls in a session so you don't have to keep specifying the wm_api_key argument
each time

## Breaking changes

- `wm_get_by_id()` changed the main argument from `x` to `ids`. This is more intuitive, since the function accepts vectors of length > 1. It is also easier to debug the code now. Help section fixed accordingly

## Minor improvements and fixes

- Fixed issue/bug #1 (errors due to deprecation of `dplyr::progress_estimated()` )
- Removed unneeded dependencies (`lwgeom`, `rlist`), add `progress` dependency due to deprecation of `dplyr::progress_estimated()`


# wikimapR 0.1.1 (2018-12-17)

- Initial public release





R/wikimapR.R
#' Import 'Wikimapia' Data as Simple Features via API
#'
#' Download and import of 'Wikimapia' data as 'sf'
#' objects.  'Wikimapia' data are extracted via the 'Wikimapia' API and
#' converted to `sf` objects retaining all recevied fields.
#'
#' @author Egor Kotov \email{kotov.egor@gmail.com}
#' @docType package
#' @name wikimapR
"_PACKAGE"


# Wrappers ----------------------------------------------------------------

safe_GET <- purrr::safely( httr::GET )

quiet_st_cast <- purrr::quietly(sf::st_cast)

# Helper functions --------------------------------------------------------

construct_polygon <- function(points_list) {

  polygon <- rbind( data.frame( x = points_list %>% purrr::map_dbl(~.x$x),
                                y = points_list %>% purrr::map_dbl(~.x$y)),
                    data.frame( x = points_list %>% purrr::map_dbl(~.x$x) %>% .[[1]],
                                y = points_list %>% purrr::map_dbl(~.x$y) %>% .[[1]])
                    ) %>%
    as.matrix(.) %>% list() %>% sf::st_polygon() %>% sf::st_sfc(crs = 4326)

  return(polygon)
}

#' Set Wikimapia API Key
#'
#' Sets an API key to make it available for all wikimapR functions calls. See details
#'
#' @param api_key Wikimapia API key
#'
#' @details
#' Use \code{set_wikimapia_api_key} to make API keys available for all Wikimapia API
#' calls in a session so you don't have to keep specifying the \code{wm_api_key} argument
#' each time.
#'
#' @references mapdeck https://github.com/SymbolixAU/mapdeck code related to API key handling was used as a template.
#'
#' @export
set_wikimapia_api_key <- function(api_key) {
  options <- api_key
  class(options) <- "wikimapia_api"
  options(wikimapia_api_key = options)
  invisible(NULL)
}






R/wm_get_by_id.R
#' Get Wikimapia Objects by ID
#'
#' Retrieves Wikimapia objects by ID.
#'
#' Uses [`place.getbyid`](http://wikimapia.org/api#placegetbyid) Wikimapia API function to retrieve full object details.
#'
#' @param ids a numeric or character vector with Wikimapia object IDs to fetch. You can get a list of objects to fetch using \link[wikimapR]{wm_get_from_bbox}
#'
#' @param language Wikimapia language to retrieve. This is specified language in ISO 639-1 format. Default language is 'en'.
#'
#' @param wm_api_key Your wikimapia API Key. If not specified, the default 'example' key is used, however it is limited to about 1 request per 30 seconds and is for testing purposes only.
#'
#' @param data_blocks The comma-separated blocks of data you want to return: main, geometry, edit, location, attached, photos, comments, translate, similar_places, nearest_places, nearest_comments, nearest_streets, nearest_hotels. Default is "main,geometry,edit,location,attached,photos,comments,translate".
#'
#' @return `sf` objects with nested details.
#'
#' @import dplyr progress sf
#'
#' @export
wm_get_by_id <- function(ids,
                         language = "en",
                         wm_api_key = getOption("wikimapia_api_key", default = "example"),
                         data_blocks = "main,geometry,edit,location,attached,photos,comments,translate"){

  pb <- progress::progress_bar$new(
    format = "[:bar] :percent",
    total = length(ids),
    clear = FALSE
  )

  wm_objects <- list()
  for(i in seq_along(ids)) {
    if( wm_api_key == "example" ) {
      Sys.sleep(30)
    } else {
      Sys.sleep(3) # the default 'up to 100 requests per 5 minutes' for Wikimapia API Keys
    }
    pb$tick()
    wm_objects[[i]] <- wikimapR:::wm_get_by_id_single(
      id = ids[[i]], language = language, wm_api_key = wm_api_key, data_blocks = data_blocks)
  }

  wm_sf_all <- wm_objects %>% purrr::map_dfr(~.x, .id = NULL) %>% sf::st_sf()

  return(wm_sf_all)
}



# non-vectorised ----------------------------------------------------------

#' Get one Wikimapia object by ID
#'
#' Retrieves single Wikimapia object by ID.
#'
#' Uses [`place.getbyid`](http://wikimapia.org/api#placegetbyid) Wikimapia API function to retrieve full object details.
#'
#' @param id a numeric or characted vector of length 1 with Wikimapia object ID to fetch. You can get a list of objects to fetch using \link[wikimapR]{wm_get_from_bbox}
#'
#' @param language Wikimapia language to retrieve. This is specified language in ISO 639-1 format. Default language is 'en'.
#'
#' @param wm_api_key Your wikimapia API Key. If not specified, the default 'example' key is used, however it is limited to about 1 request per 30 seconds and is for testing purposes only.
#'
#' @param data_blocks The comma-separated blocks of data you want to return: main, geometry, edit, location, attached, photos, comments, translate, similar_places, nearest_places, nearest_comments, nearest_streets, nearest_hotels. Default is "main,geometry,edit,location,attached,photos,comments,translate".
#'
#' @import purrr httr dplyr
#'
#' @return `sf` object with nested details.
#'
wm_get_by_id_single <- function(id,
                                language = "en",
                                wm_api_key = getOption("wikimapia_api_key", default = "example"),
                                data_blocks = "main,geometry,edit,location,attached,photos,comments,translate") {


  if ( wm_api_key == "example" ){
    warning("Using 'example' API key. This key can only be used for testing. The interval for using this key is 30 seconds. Get your own API key at http://wikimapia.org/api?action=create_key .")
  }

  base_url <- "http://api.wikimapia.org/?function=place.getbyid"

  request_url <- paste0(base_url,
                        "&id=", id,
                        "&language=", language,
                        "&key=", wm_api_key,
                        "&data_blocks=", data_blocks,
                        "&format=json")

  response <- wikimapR:::safe_GET(request_url)

  response_content <- httr::content(response$result, as = "parsed", type = "application/json", encoding = "UTF-8")

  while( any(names(response_content) %in% "debug" == TRUE) ) {
    print(response_content$debug$message)
    print("Waiting for cool down. You have probalby reached your API rate limit or using 'example' API key.")
    if( wm_api_key == "example" ) {
      Sys.sleep(30)
    } else {
      Sys.sleep(3) # the default 'up to 100 requests per 5 minutes' for Wikimapia API Keys
    }
    print("Retrying...")
    response <- wikimapR:::safe_GET(request_url)
    response_content <- httr::content(response$result, as = "parsed", type = "application/json", encoding = "UTF-8")
  }

  # process the recevied JSON

  # create polygon
  polygon <- wikimapR:::construct_polygon(response_content$polygon)

  # get flat features
  response_content_flat <- response_content[ grep("data\\.frame|list|NULL", sapply(response_content, class), invert = TRUE ) ]
  response_content_flat_df <- as.data.frame(response_content_flat, stringsAsFactors = F)

  # create an sf object
  wm_sf <- sf::st_sf(response_content_flat_df, geometry = polygon, crs = 4326)

  # add edit info
  wm_sf <- wm_sf %>% dplyr::bind_cols( response_content$edit_info %>% unlist() %>% t() %>% data.frame() )

  # add other list elements as nested lists
  response_content_nested <- response_content[ grep("data\\.frame|list|NULL", sapply(response_content, class), invert = FALSE ) ] %>%
    .[ grep("polygon|edit_info", names(.), invert = T) ]

  wm_sf_with_nested_cols <- wm_sf %>% mutate(details = list(response_content_nested))

  return(wm_sf_with_nested_cols)
}

R/subdivide_bbox.R
#' Subdivide large bounding box into smaller ones
#'
#' Subdivide large bounding box into smaller ones.
#'
#'
#'
#' @param x a numeric vector of length four, with xmin, ymin, xmax and ymax values; or a `bbox` object that is an output of \link[sf]{st_bbox}. The values must be in WGS 84 (EPSG 4326)
#'
#' @param bbox_cell_size bounding box cell size is in degrees (need to somehow fix that to work with meters across the globe). Default of 0.045 is reasonably large, roughly equivalent to 2845x5010 meters. The grid will be created in WGS84 coordinate reference system and then transformed into CRS set by `out_crs`, which is also WGS84 by default (EPSG:4326).
#'
#' @param return_bbox_or_sf object type to retun. If set to default "bbox", returns objects of type `bbox`, "sf" - returns `sf polygons` (good for plotting), "both" - returns a list of length 2 with both `bbox` and `sf` named accorgingly.
#'
#' @param out_crs EPSG code or proj4 string for the desired coordinate reference system of the output bounding box coordinates. The grid will be created in WGS84 coordinate reference system and then transformed into CRS set by `out_crs`. Default value is 4326 for WGS84.
#'
#' @import sf dplyr purrr
#'
#' @return an `sf` object grid with consequtive IDs. Also prints out a message with the number of bounding boxes and the average cell dimensions in meters.
#'
#' @export
subdivide_bbox <- function(x, bbox_cell_size = 0.045, return_bbox_or_sf = "bbox", out_crs = 4326)
{
  # x <- bbox # remove this
  # create corner coords
  corner_coords <- c( x[c(1,2)], # bottom left coord
                      x[c(1,4)], # top left coord
                      x[c(3,4)], # top right coord
                      x[c(3,2)], # bottom right coord
                      x[c(1,2)]) # bottom left coord to close the polygon

  # generate polygon from bounding box corner coordinates
  bbox_polygon <- corner_coords %>%
    matrix(ncol = 2, byrow = T) %>%
    list() %>%
    sf::st_polygon() %>%
    sf::st_sfc(crs = 4326)

  # create a grid object of type sf with consequtive IDs
  bbox_grid_sf <- sf::st_make_grid(bbox_polygon, cellsize = bbox_cell_size) %>%
    sf::st_sf() %>%
    dplyr::mutate(fid = seq(from = 1, to = nrow(.), by = 1)) %>%
    sf::st_transform(crs = out_crs)

  # get the bounding boxes of the cells in the grid object
  bbox_grid_bbox <- purrr::map(.x = bbox_grid_sf$geometry, .f = ~ sf::st_bbox(.x) )

  # evaluate approximate grid dimensions
  first_bbox_point <- bbox_grid_sf[1,] %>% wikimapR:::quiet_st_cast("POINT") %>% .$result
  size_x <- sf::st_distance(x = first_bbox_point[1,], y = first_bbox_point[2,]) %>% round(digits = 0)
  size_y <- sf::st_distance(x = first_bbox_point[2,], y = first_bbox_point[3,]) %>% round(digits = 0)

  message( paste0(nrow(bbox_grid_sf), " bounding boxes created with approximate cell size of ", size_x, "x", size_y, " meters.") )

  if (return_bbox_or_sf == "bbox") {
    return(bbox_grid_bbox)
  }

  else if (return_bbox_or_sf == "sf") {
    return(bbox_grid_sf)
  }

  else if (return_bbox_or_sf == "both") {
    bboxes <- list( bbox = bbox_grid_bbox, sf = bbox_grid_sf)
    return(bboxes)
  }
}


R/wm_get_from_bbox.R
#' Get Wikimapia Objects in a Bounding Box
#'
#' Retrieves objects in the given boundary box either with or without geometry. May also fetch just the metadata.
#'
#' Currenlty only supports bbox parameter, but not &lon_min, &lat_min, &lon_max, &lat_max and not &x, &y, &z for tiles. `Box`` endpoint is claimed to be deprecated, but the `Place.Getbyarea` does not seem to work properly.
#'
#' @param x a numeric vector of length 4, with xmin, ymin, xmax and ymax values; or a `bbox` object that is an output of \link[sf]{st_bbox}. The values must be in WGS 84 (EPSG 4326)
#'
#' @param page This is page number. 1 is default. There are usually more objects in a bounding box than the limit of 100 objects per page. Therefore you need to specify page other than 1 to get other objects. The returned list object contains number of found features and the required number of pages to retrieve them all.
#'
#' @param n_per_page This is a variable that determines the number of results per page. 100 is default (5 min, 100 max). This is `count` parameter in Wikimapia terminilogy, but different name is used here so that it does not conflict with the `dplyr::count()` function.
#'
#' @param language Wikimapia language to retrieve. This is specified language in ISO 639-1 format. Default language is 'en'.
#'
#' @param category This is wikimapia category code: category=17 - Shop, category=203 - School etc. (a detailed list will be published later) or category text query in UTF-8: category=School, category=Church etc. Default is NULL to get all categories.
#'
#' @param get_location Specifies whether to get location (centroid) and polygon geometry. You may want to disable it if you are only using the \link[wikimapR]{wm_get_from_bbox} to estimate the number of objects to reduce the returned object size.
#'
#' @param wm_api_key Your wikimapia API Key. If not specified, the default 'example' key is used, however it is limited to about 1 request per 30 seconds and is for testing purposes only.
#'
#' @param meta_only only return metadata from the response, do not return feature attributes or geometry.
#'
#' @return If `meta_only = FALSE` returns a list with data.frame of object attributes (and object geometry if `get_location = TRUE`). If `meta_only = TRUE` only returns metadata of the responce: the number of objects in the bounding box (`found`), version, language, current page, current `n_per_page`.
#' If `get_location = TRUE` returns a list object with `sf polygons` with all object attributes, `sf points` with all object attributes (the centroids), and also the metadata described above.
#'
#' @import purrr httr dplyr sf
#'
#' @export
wm_get_from_bbox <- function(x,
                             page = 1,
                             n_per_page = 100,
                             language = "en",
                             category = NULL,
                             get_location = TRUE,
                             wm_api_key = getOption("wikimapia_api_key", default = "example"),
                             meta_only = FALSE){

  if ( wm_api_key == "example" ){
    warning("Using 'example' API key. This key can only be used for testing. The interval for using this key is 30 seconds. Get your own API key at http://wikimapia.org/api?action=create_key .")
  }

  bbox_string <- paste(x, collapse = ",")

  base_url <- "http://api.wikimapia.org/?function=box"

  request_url <- paste0(base_url,
                        "&page=", page,
                        "&count=", n_per_page,
                        "&language=", language,
                        if(is.null(category) == FALSE) { paste0("&category=", category)  } ,
                        "&bbox=", bbox_string,
                        if(get_location == FALSE) { "&disable=location,polygon" },
                        "&key=", wm_api_key,
                        "&format=json")



  response <- wikimapR:::safe_GET(request_url)

  response_content <- httr::content(response$result, as = "parsed", type = "application/json", encoding = "UTF-8")

  while( any(names(response_content) %in% "debug" == TRUE) ) {
    print(response_content$debug$message)
    print("Waiting for cool down. You have probalby reached your API rate limit or using 'example' API key.")
    if( wm_api_key == "example" ) {
      Sys.sleep(30)
    } else {
      Sys.sleep(3) # the default 'up to 100 requests per 5 minutes' for Wikimapia API Keys
    }
    print("Retrying...")
    response <- wikimapR:::safe_GET(request_url)
    response_content <- httr::content(response$result, as = "parsed", type = "application/json", encoding = "UTF-8")
  }

  meta_df <- response_content[c("version", "language", "page", "count", "found")] %>% as.data.frame(stringsAsFactors = FALSE)
  meta_df$found <- as.integer(meta_df$found)

  if ( meta_only == TRUE ) {
    return(meta_df)
  }

  # process the received JSON

  ## get primary data about objects
  wm_objects_attributes <- data.frame(id = response_content$folder %>% purrr::map_chr(~ .x$id),
                                      name = response_content$folder %>% purrr::map_chr(~ .x$name),
                                      url = response_content$folder %>% purrr::map_chr(~ .x$url),
                                      stringsAsFactors = FALSE)

  if ( get_location == FALSE ){
    wm <- list(df = wm_objects_attributes,
               meta = meta_df)
    return(wm)
  }

  ## create polygons and centroids from raw JSON data
  if (  get_location == TRUE ) {

    # create polygons
    polygons <- response_content$folder %>%
      purrr::map(~ wikimapR:::construct_polygon(.x$polygon) ) %>%
      do.call(c, .) %>%
      sf::st_sf(geometry = .) %>%
      dplyr::mutate(id = wm_objects_attributes$id) %>%
      merge(x = ., y = wm_objects_attributes, by = "id", all = T)


    # create centroids
    centroids <- response_content$folder %>%
      purrr::map(~ c( x = .x$location$lon,
                      y = .x$location$lat)) %>%
      purrr::map(~ st_point(.) %>% sf::st_sfc(crs = 4326) ) %>%
      do.call(c, .) %>%
      sf::st_sf(geometry = ., id = wm_objects_attributes$id) %>%
      merge(x = ., y = wm_objects_attributes, by = "id", all = T)

    wm <- list(polygons = polygons,
               centroids = centroids,
               meta = meta_df)
    return(wm)
  }
}
